### PR TITLE
add support for range based File reading/writing APIs

### PR DESCRIPTION
**Pull request by rnjtranjan**

### CODE REVIEW

**Comment by rnjtranjan** [comment]
fix #662

File: src/Streamly/Internal/FileSystem/Handle.hs [hunk 1]
```diff
@@ -100,13 +102,21 @@ module Streamly.Internal.FileSystem.Handle
...
 import Control.Monad.IO.Class (MonadIO(..))
+import Data.IORef
 import Data.Word (Word8)
 import Foreign.ForeignPtr (withForeignPtr)
 import Foreign.ForeignPtr.Unsafe (unsafeForeignPtrToPtr)
 import Foreign.Ptr (minusPtr, plusPtr)
 import Foreign.Storable (Storable(..))
 import GHC.ForeignPtr (mallocPlainForeignPtrBytes)
-import System.IO (Handle, hGetBufSome, hPutBuf, stdin, stdout)
+import System.IO ( Handle
```
Comment (adithyaov) [inline src/Streamly/Internal/FileSystem/Handle.hs:112]: Maybe change this to,
```
import System.IO
       (Handle, SeekMode(..), hGetBufSome, hPutBuf, hSeek, stdin, stdout)
```

File: src/Streamly/Internal/FileSystem/Handle.hs [hunk 2]
```diff
@@ -481,6 +491,65 @@ write = writeWithBufferOf defaultChunkSize
...
 
+
+toChunksWithBufferOfRange
+    :: (IsStream t, MonadIO m) 
+    => Int 
+    -> IORef Int 
+    -> Handle 
+    -> t m (Array Word8)
+toChunksWithBufferOfRange size lenRef h = D.fromStreamD (D.Stream step ())
+  where
```
Comment (adithyaov) [inline src/Streamly/Internal/FileSystem/Handle.hs:502]: Indent `where` by 4 and add a new line.
So something like,
```
toChunksWithBufferOfRange size lenRef h = D.fromStreamD (D.Stream step ())
    where

    {-# INLINE_LATE step #-}
    ...
```

File: src/Streamly/Internal/FileSystem/Handle.hs [hunk 3]
```diff
@@ -481,6 +491,65 @@ write = writeWithBufferOf defaultChunkSize
 write2 :: MonadIO m => Fold2 m Handle Word8 ()
 write2 = writeWithBufferOf2 defaultChunkSize
 
+
+toChunksWithBufferOfRange
```
Comment (adithyaov) [inline src/Streamly/Internal/FileSystem/Handle.hs:495]: I think you've missed an inline pragma here

File: src/Streamly/Internal/FileSystem/Handle.hs [hunk 4]
```diff
@@ -481,6 +491,65 @@ write = writeWithBufferOf defaultChunkSize
...
+{-# INLINABLE writeArrayWith #-}
+writeArrayWith :: Storable a => Handle -> Array a -> Int -> IO ()
+writeArrayWith _ arr _ | A.length arr == 0 = return ()
+writeArrayWith h Array{..} len = withForeignPtr aStart $ \p -> hPutBuf h p len  
+
+{-# INLINE fromChunksWith #-}
+fromChunksWith :: (MonadIO m, Storable a)
+    => Handle -> IORef Int -> SerialT m (Array a) -> m ()
+fromChunksWith h lenRef = do    
+    S.mapM_ (\arr -> do
```
Comment (adithyaov) [inline src/Streamly/Internal/FileSystem/Handle.hs:530]: You can probably take this function out into a where clause and rewrite this as `S.mapM_ func`

File: src/Streamly/Internal/FileSystem/Handle.hs [hunk 5]
```diff
@@ -481,6 +491,65 @@ write = writeWithBufferOf defaultChunkSize
...
+
+{-# INLINABLE writeArrayWith #-}
+writeArrayWith :: Storable a => Handle -> Array a -> Int -> IO ()
+writeArrayWith _ arr _ | A.length arr == 0 = return ()
+writeArrayWith h Array{..} len = withForeignPtr aStart $ \p -> hPutBuf h p len  
+
+{-# INLINE fromChunksWith #-}
+fromChunksWith :: (MonadIO m, Storable a)
+    => Handle -> IORef Int -> SerialT m (Array a) -> m ()
+fromChunksWith h lenRef = do
```
Comment (adithyaov) [inline src/Streamly/Internal/FileSystem/Handle.hs:529]: I think you can omit this `do` keyword.

File: src/Streamly/Internal/FileSystem/Handle.hs [hunk 6]
```diff
@@ -526,14 +521,16 @@ writeArrayWith h Array{..} len = withForeignPtr aStart $ \p -> hPutBuf h p len
 {-# INLINE fromChunksWith #-}
 fromChunksWith :: (MonadIO m, Storable a)
     => Handle -> IORef Int -> SerialT m (Array a) -> m ()
-fromChunksWith h lenRef = do    
-    S.mapM_ (\arr -> do 
+fromChunksWith h lenRef = S.mapM_ func
```
Comment (adithyaov) [inline src/Streamly/Internal/FileSystem/Handle.hs:524]: Maybe rename `func` to something more informative?

File: src/Streamly/Internal/FileSystem/Handle.hs [hunk 7]
```diff
@@ -526,14 +521,16 @@ writeArrayWith h Array{..} len = withForeignPtr aStart $ \p -> hPutBuf h p len
...
+    
+    where
+    
+    func = \arr -> do
```
Comment (adithyaov) [inline src/Streamly/Internal/FileSystem/Handle.hs:528]: You can replace this with `func arr`

**Comment by harendra-kumar** [comment]
Rebased.

**Comment by harendra-kumar** [comment]
We need to implement only unfolds for this. Please implement the signature of unfold that I added in #1189.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: src/Streamly/Internal/FileSystem/Handle.hs [hunk 8]
```diff
@@ -257,9 +257,34 @@ readChunksWithBufferOf = Unfold step return
 --
 --
 {-# INLINE_NORMAL readChunksFromToWith #-}
-readChunksFromToWith :: -- MonadIO m =>
+readChunksFromToWith :: MonadIO m =>
     Unfold m (Int, Int, Int, Handle) (Array Word8)
```
[the remark this answers is not in the document]
Comment (rnjtranjan) [reply src/Streamly/Internal/FileSystem/Handle.hs:262]: It goes beyond 80 columns.

File: src/Streamly/Internal/FileSystem/Handle.hs [hunk 9]
```diff
@@ -257,9 +258,34 @@ readChunksWithBufferOf = Unfold step return
...
+    where
+
+    inject (from, to, buffSize, h) = do
+        liftIO $ hSeek h AbsoluteSeek $ fromIntegral from
+        return (to - from + 1, buffSize, h)
+
+    {-# INLINE_LATE step #-}
+    step (remaining, buffSize, h) =
+        if remaining <= 0
+        then return D.Stop -- if range is unsafe return empty stream
```
Comment (harendra-kumar) [inline src/Streamly/Internal/FileSystem/Handle.hs:274]: The code is obvious so we can remove the comment.

File: src/Streamly/Internal/FileSystem/Handle.hs [hunk 10]
```diff
@@ -257,9 +258,34 @@ readChunksWithBufferOf = Unfold step return
...
+        else
+            do
+            arr <- getChunk buffSize h
```
Comment (harendra-kumar) [inline src/Streamly/Internal/FileSystem/Handle.hs:277]: We can make this `getChunk (min bufSize remaining) h`.

File: src/Streamly/Internal/FileSystem/Handle.hs [hunk 11]
```diff
@@ -257,9 +258,34 @@ readChunksWithBufferOf = Unfold step return
...
+                case A.length arr of
+                    0 -> D.Stop
+                    len ->
+                        if (remaining - len) >= 0
+                        then
+                            D.Yield arr (remaining - len, buffSize, h)
+                        else
+                            D.Yield
+                            (A.unsafeSlice 0 remaining arr)
+                            (0, 0, h)
```
Comment (harendra-kumar) [inline src/Streamly/Internal/FileSystem/Handle.hs:288]: If we use `min` in `getChunk` then this should not occur. We can put a `error "Bug: ...` message here.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: src/Streamly/Internal/FileSystem/Handle.hs [hunk 12]
```diff
@@ -257,9 +257,32 @@ readChunksWithBufferOf = Unfold step return
...
+            arr <- getChunk (min buffSize remaining) h
+            return $
+                case A.length arr of
+                    0 -> D.Stop
+                    len ->
+                        if (remaining - len) >= 0
+                        then
+                            D.Yield arr (remaining - len, buffSize, h)
+                        else
+                            error "Bug: ..."
```
Comment (harendra-kumar) [inline src/Streamly/Internal/FileSystem/Handle.hs:285]: I meant some message in place of `...`, e.g. ` error "Bug: readFromChunksFromToWith: array length more than expected`.

**Review by harendra-kumar — COMMENTED** [review]
We can add a test as well to establish the basic sanity of the API.

**Review by harendra-kumar — COMMENTED** [review]
All these tests are very similar. It should be possible to factor out the common code in a single function and pass different args to it for different test cases.

**Review by rnjtranjan — COMMENTED** [review]
Looks good.

**Pull request merged**